File: src/utils/vocabIndex.js
import { PHASES } from '../data/studyData';

// A leading Italian article (definite/indefinite) so that a vocab entry like
// "il Verbo" or "la luce" is also reachable by its bare content word ("verbo",
// "luce") — which is how the word appears inside an example sentence.
const LEADING_ARTICLE = /^(l['’]|gli\s+|le\s+|il\s+|lo\s+|la\s+|i\s+|uno\s+|una\s+|un['’]?\s*)/i;

function clean(s) {
  return s.toLowerCase().trim();
}

function stripArticle(s) {
  return s.replace(LEADING_ARTICLE, '').trim();
}

let INDEX = null;

// Build once: maps a normalized Italian term → { it, en, ipa }. Both the full
// term and its article-stripped form are indexed. First write wins, so the
// earliest (lowest-week) gloss is the canonical one for a repeated word.
export function getVocabIndex() {
  if (INDEX) return INDEX;
  INDEX = new Map();
  const add = (key, val) => {
    const k = clean(key);
    if (k && !INDEX.has(k)) INDEX.set(k, val);
  };
  for (const phase of PHASES) {
    for (const week of phase.weeks) {
      for (const [it, en, , ipa] of week.vocab) {
        const val = { it, en, ipa: ipa || '' };
        add(it, val);
        const stripped = stripArticle(clean(it));
        if (stripped && stripped !== clean(it)) add(stripped, val);
      }
    }
  }
  return INDEX;
}

// Look up a single word/token. Returns { it, en, ipa } or null. Tries the word
// as-is, then with any leading article stripped (so "l'unzione" and "unzione"
// both resolve).
export function lookupWord(word) {
  if (!word) return null;
  const idx = getVocabIndex();
  const w = clean(word);
  return idx.get(w) || idx.get(stripArticle(w)) || null;
}

// Split an Italian string into word / non-word tokens, preserving every
// character so the original text can be reconstructed exactly. A "word" keeps
// internal apostrophes (l'unzione, dov'è) as a single token.
export function tokenize(text) {
  const tokens = [];
  const re = /[A-Za-zÀ-ÿ]+(?:['’][A-Za-zÀ-ÿ]+)*/g;
  let last = 0;
  let m;
  while ((m = re.exec(text)) !== null) {
    if (m.index > last) tokens.push({ text: text.slice(last, m.index), isWord: false });
    tokens.push({ text: m[0], isWord: true });
    last = m.index + m[0].length;
  }
  if (last < text.length) tokens.push({ text: text.slice(last), isWord: false });
  return tokens;
}

File: ios-native/BibbiaCore/Sources/BibbiaCore/VocabIndex.swift
import Foundation

// Port of src/utils/vocabIndex.js — O(1) lookup of any Italian word to its
// vocab gloss ({ it, en, ipa }), reachable by the full term and its
// article-stripped stem (so "il Verbo" is found as "verbo"), plus the
// word/non-word tokenizer that preserves the original text exactly.

public struct VocabGloss: Equatable {
    public let it: String
    public let en: String
    public let ipa: String
}

public final class VocabIndex {
    private var index: [String: VocabGloss] = [:]
    private let articles: [String]

    /// Build once from the course. First write wins, so the earliest
    /// (lowest-week) gloss is the canonical one for a repeated word.
    public init(course: Course) {
        self.articles = course.locale.articles
        for phase in course.phases {
            for week in phase.weeks {
                for card in week.vocab {
                    let gloss = VocabGloss(it: card.it, en: card.en, ipa: card.ipa ?? "")
                    add(card.it, gloss)
                    let cleaned = clean(card.it)
                    let stripped = stripArticle(cleaned)
                    if !stripped.isEmpty, stripped != cleaned {
                        add(stripped, gloss)
                    }
                }
            }
        }
    }

    private func clean(_ s: String) -> String {
        s.lowercased().trimmingCharacters(in: .whitespacesAndNewlines)
    }

    private func stripArticle(_ s: String) -> String {
        Articles.stripLeading(s, articles: articles)
            .trimmingCharacters(in: .whitespacesAndNewlines)
    }

    private func add(_ key: String, _ val: VocabGloss) {
        let k = clean(key)
        if !k.isEmpty, index[k] == nil { index[k] = val }
    }

    /// Look up a single word/token. Tries the word as-is, then with any
    /// leading article stripped (so "l'unzione" and "unzione" both resolve).
    public func lookup(_ word: String) -> VocabGloss? {
        if word.isEmpty { return nil }
        let w = clean(word)
        return index[w] ?? index[stripArticle(w)]
    }

    public var count: Int { index.count }
}

public struct TextToken: Equatable {
    public let text: String
    public let isWord: Bool
}

// Mirrors the JS tokenizer's [A-Za-zÀ-ÿ] word-character class.
private func isWordChar(_ c: Character) -> Bool {
    guard let scalar = c.unicodeScalars.first, c.unicodeScalars.count == 1 else { return false }
    let v = scalar.value
    return (0x41...0x5A).contains(v) || (0x61...0x7A).contains(v) || (0xC0...0xFF).contains(v)
}

/// Split an Italian string into word / non-word tokens, preserving every
/// character so the original text can be reconstructed exactly. A "word"
/// keeps internal apostrophes (l'unzione, dov'è) as a single token.
public func tokenize(_ text: String) -> [TextToken] {
    var tokens: [TextToken] = []
    let chars = Array(text)
    var i = 0
    var pendingStart = 0

    while i < chars.count {
        if isWordChar(chars[i]) {
            if i > pendingStart {
                tokens.append(TextToken(text: String(chars[pendingStart..<i]), isWord: false))
            }
            var end = i
            while end < chars.count {
                if isWordChar(chars[end]) {
                    end += 1
                } else if (chars[end] == "'" || chars[end] == "’"),
                          end + 1 < chars.count, isWordChar(chars[end + 1]) {
                    end += 1  // internal apostrophe followed by a letter
                } else {
                    break
                }
            }
            tokens.append(TextToken(text: String(chars[i..<end]), isWord: true))
            i = end
            pendingStart = end
        } else {
            i += 1
        }
    }
    if pendingStart < chars.count {
        tokens.append(TextToken(text: String(chars[pendingStart...]), isWord: false))
    }
    return tokens
}
